packages/typespec-vscode/src/openapi3-preview.ts

```diff
@@ -0,0 +1,251 @@
+  });
+
+  switch (files.length) {
+    case 0:
+      const errMsg = "No 'main.tsp' file found in the workspace.";
+      logger.error(errMsg);
+      vscode.window.showErrorMessage(errMsg);
+      return undefined;
+    case 1:
+      return files[0].fsPath;
+    default:
+      return await vscode.window.showQuickPick(
+        files.map((file) => file.fsPath),
+        {
+          title: "Select the 'main.tsp' file",
+        },
+      );
+  }
+}
+
+const openApi3TempFolders = new Map<string, string>();
+const openApi3PreviewPanels = new Map<string, vscode.WebviewPanel>();
+const selectedOpenApi3OutputFiles = new Map<string, string>();
+
+export async function loadOpenApi3PreviewPanel(
+  mainTspFile: string,
+  context: vscode.ExtensionContext,
+  client: TspLanguageClient,
+) {
+  const compilerVersion = client.initializeResult?.serverInfo?.version;
+  if (!compilerVersion || semver.lt(compilerVersion, "0.65.0")) {
+    const errMsg = "OpenAPI3 preview requires TypeSpec compiler version 0.65.0 or later.";
+    logger.info(errMsg);
+    vscode.window.showErrorMessage(errMsg);
+    return;
+  }
+
+  if (openApi3PreviewPanels.has(mainTspFile)) {
+    openApi3PreviewPanels.get(mainTspFile)!.reveal();
+  } else {
+    const getOpenApi3Output = async (uri?: vscode.Uri): Promise<string | undefined> => {
+      return await vscode.window.withProgress(
+        {
+          location: vscode.ProgressLocation.Notification,
+          title: "Loading OpenAPI3 files...",
+        },
+        async (): Promise<string | undefined> => {
+          const srcFolder = dirname(mainTspFile);
+          const outputFolder = await getOutputFolder(mainTspFile);
+          let result: ExecOutput | undefined;
+          try {
+            result = await client.compileOpenApi3(mainTspFile, srcFolder, outputFolder);
+          } catch (e) {
+            const error = e as ExecOutput;
+            const errMsg = `Failed to generate OpenAPI3 files: ${error.stderr || error.stdout || error.error}`;
+            logger.error(errMsg);
+            vscode.window.showErrorMessage(errMsg);
+            return;
+          }
+          if (result === undefined || result.exitCode !== 0) {
+            const errMsg = result?.stderr ?? "Failed to generate OpenAPI3 files.";
+            logger.error(errMsg);
+            vscode.window.showErrorMessage(errMsg);
+            return;
+          } else {
+            const outputs = await readdir(outputFolder);
+            if (outputs.length === 0) {
+              const errMsg = result?.stderr ?? "No openAPI3 file generated.";
+              logger.error(errMsg);
+              vscode.window.showErrorMessage(errMsg);
+              return;
+            } else if (outputs.length === 1) {
+              const first = outputs[0];
+              return parseOpenApi3File(join(outputFolder, first));
+            } else {
+              if (selectedOpenApi3OutputFiles.has(mainTspFile)) {
```

Comment:
what if user wants to select another one to preview?

Reply:
change to prompt users to select every time the preview is manually triggered